drivers/disk/flashdisk.c

```diff
@@ -184,35 +228,42 @@ static int update_flash_block(off_t start_addr, uint32_t size, const void *buff)
 static int disk_flash_access_write(struct disk_info *disk, const uint8_t *buff,
 				 uint32_t start_sector, uint32_t sector_count)
 {
+	struct flashdisk_data *ctx;
 	off_t fl_addr;
 	uint32_t remaining;
 	uint32_t size;
 
-	fl_addr = lba_to_address(start_sector);
+	ctx = CONTAINER_OF(disk, struct flashdisk_data, info);
+
+	fl_addr = ctx->offset + start_sector * SECTOR_SIZE;
+	if (fl_addr >= (ctx->offset + ctx->size)) {
+		LOG_ERR("Address %lx outside partition boundary", (long)fl_addr);
+		return -EINVAL;
+	}
+
 	remaining = (sector_count * SECTOR_SIZE);
 
 	/* check if start address is erased-aligned address  */
-	if (fl_addr & (CONFIG_DISK_FLASH_ERASE_ALIGNMENT - 1)) {
+	if (fl_addr & (ctx->page_size - 1)) {
 		off_t block_bnd;
 
 		/* not aligned */
 		/* check if the size goes over flash block boundary */
-		block_bnd = fl_addr + CONFIG_DISK_ERASE_BLOCK_SIZE;
-		block_bnd = block_bnd & ~(CONFIG_DISK_ERASE_BLOCK_SIZE - 1);
+		block_bnd = fl_addr + ctx->page_size;
+		block_bnd = block_bnd & ~(ctx->page_size - 1);
 		if ((fl_addr + remaining) < block_bnd) {
```

Comment:
I think the condition here should be `<=` instead of `<`.

For example Page Size 4096, Sector size 512, Write start sector 7, sector count 1, partition offset (`ctx->offset`) 0 means that 512 bytes are written at address `7*512=3584`. The bytes written are at offsets `[3584; 3584 + 512 - 1]`. `block_bnd` would be calculated as `(3584+512) & ~(512 - 1) = 4096` and `fl_addr + remaining` would be `3584 + 512 = 4096`.

Hopefully the use of `<` has no actual impact, as the `while (remaining)` loops below would simply be skipped due to `remaining` being `0` on entry.